lib/loki-logproto/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+pub mod stats {
+    include!(concat!(env!("OUT_DIR"), "/stats.rs"));
+}
+
+pub mod logproto {
+    include!(concat!(env!("OUT_DIR"), "/logproto.rs"));
+}
+
+pub mod util {
+    use super::logproto;
+    use prost::Message;
+    use std::collections::HashMap;
+
+    const NANOS_RANGE: i64 = 1_000_000_000;
+
+    // (<Timestamp in nanos>, <Line>)
+    pub struct Entry(pub i64, pub String);
+
+    impl From<Entry> for logproto::EntryAdapter {
+        fn from(entry: Entry) -> Self {
+            logproto::EntryAdapter {
+                timestamp: Some(prost_types::Timestamp {
+                    seconds: entry.0 / NANOS_RANGE,
+                    nanos: (entry.0 % NANOS_RANGE) as i32,
+                }),
+                line: entry.1,
+            }
+        }
+    }
+
+    // (<Labels>, <Lines>)
+    pub struct Batch(pub HashMap<String, String>, pub Vec<Entry>);
+
+    impl From<Batch> for logproto::StreamAdapter {
+        fn from(batch: Batch) -> Self {
+            let labels = encode_labels_map_to_string(&batch.0);
+            let entries: Vec<logproto::EntryAdapter> =
+                batch.1.into_iter().map(|entry| entry.into()).collect();
+
+            logproto::StreamAdapter {
+                labels,
+                entries,
+                hash: 0,
+            }
+        }
+    }
+
+    impl Batch {
+        pub fn encode(self) -> Vec<u8> {
+            let stream: logproto::StreamAdapter = self.into();
+            let push_request = logproto::PushRequest {
+                streams: vec![stream],
+            };
+            let buf = push_request.encode_to_vec();
+            let mut encoder = snap::raw::Encoder::new();
+            encoder.compress_vec(&buf).expect("out of memory")
+        }
+    }
+
+    const RESERVED_LABEL_TENANT_ID: &str = "__tenant_id__";
+    const RESERVED_LABELS: [&str; 1] = [RESERVED_LABEL_TENANT_ID];
+
+    // ref: https://github.com/grafana/loki/blob/65c6e254bd22151ab7fc84ec46e13eee2e354aa0/clients/pkg/promtail/client/batch.go#L61-L75
+    pub fn encode_labels_map_to_string(labels: &HashMap<String, String>) -> String {
+        let mut labels: Vec<String> = labels
+            .iter()
+            .filter(|(k, _)| !RESERVED_LABELS.contains(&k.as_str()))
+            .map(|(k, v)| format!("{}={:?}", k, v))
```

Comment:
Why use the debug formatter for the value? It's already a string.

Reply:
Hi, Loki requires the label value of kv-pair to be formatted into double-quoted string, the [corresponding implementation](https://github.com/grafana/loki/blob/main/clients/pkg/promtail/client/batch.go#L70) in Promtail shows:

```
		lstrs = append(lstrs, fmt.Sprintf("%s=%q", l, v))
```

That's why I choose debug formatter here, is that the preferable way to format string into double-quoted style in Rust?

Reply:
Ah, gotcha.

The preferable way would be to actually specify the double quotes, like so:


```rust
format!("{}=\"{}\"", k, v)
```

Reply:
Aha, that's quite easy! I have pushed my latest commit to improve it. Thanks for making a review for that!